### ISSUE TITLE
Signing the execution cost and the redeemer script 

### ISSUE THREAD

**Issue by polinavino** [input-output-hk/cardano-ledger-specs#1057]
Who/how/where should or even can sign these things? 

Both of these things need signatures or some other proof that they have not been tampered with by the slot leader.

**Comment by polinavino** [comment]
My concern is that there are no signatures on these things that are done by the parties interested in these pieces of data being correct and untampered with. This is true even if they are both stored inside the body of the transaction, which is signed by numerous keys, but not by the scripts in any meaningful way. Am I missing something here?

**Comment by polinavino** [comment]
In the case of the inputs marked for-fees, the keys associated with the addresses they are spending from already have to sign the transaction, and that is enough to guarantee that the execution cost and the redeemer are in effect signed by the parties interested in things validating. 

So I think this is actually not a big concern in terms of malicious parties tampering with the fees or substituting a redeemer that results in non-validation of a script.

**Comment by michaelpj** [comment]
Let me try and summarise my understanding of yesterday's discussion:

Confusingly, pubkey signatures on transaction bodies are *not* about transaction integrity (although this is a common use case for signatures). Instead, they are *only* there to authorize spending from certain outputs. To see this, consider that someone can already tamper with a transaction in flight to remove a signature, which won't be noticeable and will prevent it from validating. So that's not an attack we are (currently) trying to prevent.

So the key determinant of "does this go in the transaction body" is not "does this need to be signed" but "is this a determinant of how the money moves". Things which prove that the transaction is allowed to be added to the chain, but which do not affect what it *does*, do not need to be in the body.

This means that everything related to scripts doesn't need to be in the body. That means validators, redeemers, execution costs, etc. *Something* that determines the overall fees *does* need to be in the body, since this affects where some money goes. But this could just be an overall fee field (see https://github.com/input-output-hk/cardano-ledger-specs/issues/1047#issuecomment-559427511 for some consequences).

So basically, the whole integrity thing is a red herring and I apologize for muddying the waters!

**Comment by polinavino** [comment]
I think this makes sense, we only need to worry that all money movement is authorized. I still think the redeemer needs to be inside the body so that it is signed by the parties that will be charged fees in case of script validation failure.

**Comment by michaelpj** [comment]
> I still think the redeemer needs to be inside the body so that it is signed by the parties that will be charged fees in case of script validation failure.

Yeeeeeees... It's an awkward argument, though. Suddenly we do care about integrity, and we're trying to piggback on our existing signature mechanism (which doesn't try and ensure integrity) to get it. We need to write out the argument carefully, I think it's something like:
- Suppose X has signed a transaction T which also includes some script inputs.
- I attempt to intercept this and make script validation fail by altering the redeemers. Then either:
    - I do not remove the signature and the signature will prevent me from altering the body.
    - I remove the signature (which I can do), and then either:
        - Some input fails validation at stage 1
        - None of X's money was at risk anyway.

**Comment by polinavino** [comment]
It is a bit piggy-backy, but still valid, I guess? 
We should add that if any money belonging to X is moved by a processed transaction T (whether all scripts validate or not), X will be forced to sign off on it.

**Comment by polinavino** [comment]
Hash of redeemer structure is inside the body, execution budget is also inside the body. Body is hashed and the hash is signed.
